### DISCUSSION TITLE
How to Implement Page Caching Using Server Middleware in Nuxt3

### DISCUSSION THREAD

**Discussion by Wulawa** [nuxt/nuxt#26957] [Questions]
I am currently working on a Nuxt.js project and I am trying to increase the performance of my application by implementing page caching. I would like to manually implement this within a server middleware in Nuxt.js.
﻿
Could you please provide some guidance on how to achieve this? I am particularly interested in understanding:
﻿

In nuxtjs2, I will use `res.end()` to respond to caching in advance in serverMiddleware, but in nuxt3, it is not recommended to respond to requests in Server Middleware

I appreciate any help or resources you can provide on this topic.
﻿
Thank you for your time.

**Comment by TheAlexLichter** [comment]
You can use route rules for now. For a more fine-grained approach, follow https://github.com/nuxt/nuxt/issues/26598

**Comment by xjccc** [comment]
Like manniL said, you could look at this.
[routeRules](https://nuxt.com/docs/api/nuxt-config#routerules-1)

or

[Legacy Handler or Middleware](https://nuxt.com/docs/guide/directory-structure/server#legacy-handler-or-middleware)

**Reply by Wulawa** [reply]
Thank you very much for your suggestion. Yes, using fromNodeMiddleware is work, Currently, I have implemented it in the following way:
```javascript
export default defineEventHandler(async (event) => {
	const cacheableKey  = await getCacheableKey(event);
	if (cacheableKey) {
		const hit = cachePage.get(cacheableKey);
		return hit
	}
	event.node.res.original_end = event.node.res.end;
	event.node.res.end  =  function (html) {
		if (event.node.res.statusCode === 200) {
			cachePages.set(cacheableKey, html);
		}
		event.node.res.original_end(html)
		return this;
	}
})
```
Do you see any issues with this approach?

**Reply by xjccc** [reply]
😂 For long time ago, i have a similar scene.
[cache](https://github.com/xjccc/nuxt-error-and-cache/blob/bd583d790322e601848a3ce3dde8dabd8fc2abb6/src/runtime/cache.ts#L54)
But i don't think its a good complete. For now we could use `routeRules`

**Reply by Wulawa** [reply]
Ok, Thank you once again. I'll try `routeRules `. Also, do you have any good examples of `routeRules` that you could share?😊

**Reply by xjccc** [reply]
Look at [routeRules](https://nitro.unjs.io/config#routerules) for some examples in your scene.
